hugr-core/src/extension/resolution/types.rs

```diff
@@ -1,174 +1,214 @@
-//! Resolve weak links inside `CustomType`s in an optype's signature.
+//! Collect the extensions referenced inside `CustomType`s in an optype or
+//! signature.
+//!
+//! Fails if any of the weak extension pointers have been invalidated.
+//!
+//! See [`super::update_op_types_extensions`] for a mutating version that
+//! updates the weak links to point to the correct extensions.
 
-use std::sync::Arc;
+use std::collections::HashSet;
 
-use super::{ExtensionRegistry, ExtensionResolutionError};
-use crate::ops::OpType;
+use super::ExtensionCollectionError;
+use crate::extension::{ExtensionId, ExtensionRegistry};
+use crate::ops::{DataflowOpTrait, OpType};
 use crate::types::type_row::TypeRowBase;
-use crate::types::{MaybeRV, Signature, SumType, TypeBase, TypeEnum};
+use crate::types::{FuncTypeBase, MaybeRV, SumType, TypeArg, TypeBase, TypeEnum};
 use crate::Node;
 
-/// Replace the dangling extension pointer in the [`CustomType`]s inside a
-/// signature with a valid pointer to the extension in the `extensions`
-/// registry.
+/// Collects every extension used te define the types in an operation.
 ///
-/// When a pointer is replaced, the extension is added to the
-/// `used_extensions` registry and the new type definition is returned.
+/// Custom types store a [`Weak`] reference to their extension, which can be
+/// invalidated if the original `Arc<Extension>` is dropped. This normally
+/// happens when deserializing a HUGR. On such cases, we return an error with
+/// the missing extension names.
 ///
-/// This is a helper function used right after deserializing a Hugr.
-pub fn update_op_types_extensions(
-    node: Node,
-    op: &mut OpType,
-    extensions: &ExtensionRegistry,
-    used_extensions: &mut ExtensionRegistry,
-) -> Result<(), ExtensionResolutionError> {
+/// Use [`collect_op_types_extensions`] instead to update the weak references and
```

Comment:
Think this should point to `update_` something, rather than (AFAICS) the method this doc comment is on?